### PR TITLE
PHOENIX-7006: Configure maxLookbackAge at table level 

### PR DESCRIPTION
**Pull request by sanjeet006py**
[Design doc](https://docs.google.com/document/d/12q_ujY1FwftCkQGbeSzFrRJRGy4Vb-JRbK_3N-XiDzU/edit)

- Support for table level max lookback age in SYSCAT.
- Support for table level max lookback age in CREATE/ALTER TABLE.
- Block setting/altering max lookback age for views and indexes.
- Retrieve max lookback age for view and index from data table via hierarchical lookup and store it only for data table.
- Invalidate metadata cache on alter table for child views and associated indexes (view indexes and data table indexes) so that new PTable instance in loaded in cache with update max lookback age.
- Use table level max lookback age in coproc hooks wherever cluster level max lookback age is being used. Cluster level max lookback age will be fallback if table level max lookback age is not set.
- Use table level max lookback in Index verification in GlobalIndexRegionScanner. Cluster level max lookback age will be fallback if table level max lookback age is not set.
- Use table level max lookback age in IndexScrutinyTool.
- Test coverage for all the above items.

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java:1713]: nit: 259200000 can be a variable or constant since it's used twice in this method.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java:3161]: When creating a table from now on maxlookbackage will default to 0 instead of NULL? Are we sure that we aren't introducing assumptions that max lookback age is non null elsewhere, since existing tables will have the value be null?

File: phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java:5916]: So here max lookback age is a Long, and we have some logic dealing with it being null at, e.g line 5578...but elsewhere like on PTable we're assuming it's a long that can't be null.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/PTable.java
Comment (gjacoby126) [inline phoenix-core/src/main/java/org/apache/phoenix/schema/PTable.java:1001]: We're absolutely sure that every path that can possibly create a PTable from a table without a max lookback age will transparently use the default?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (lokiore) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1835]: Can you also add test in which you are asserting maxlookback through PTable object, I mean getting maxlookback from getMaxLookbackAge() API so that, that code path is also tested. 
Also add assertions in some old tests or new one where value is null in SYSCAT.

**Comment by tkhurana** [comment]
Another thing we should consider is that maxlookback on index should inherit the value of the base table. The behavior should be exactly similar to TTL where both the data table and index have same values and you cannot directly set ttl on index. It always inherits the TTL value from base table.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java
Comment (sanjeet006py) [reply phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java:3161]: Got it, so I think the max lookback age as undefined should be either null or 0. And, preferrable will be null as already for existing tables we will have it as NULL. Please correct me if I am understanding it wrong.

**Comment by sanjeet006py** [comment]
> Another thing we should consider is that maxlookback on index should inherit the value of the base table. The behavior should be exactly similar to TTL where both the data table and index have same values and you cannot directly set ttl on index. It always inherits the TTL value from base table.

Yes, I will be doing that only. Current set of changes include SYSCAT schema changes and CREATE/ALTER Table support. In the subsequent changes I will be adding support to block max lookback age modifications for index.

File: phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java
Comment (sanjeet006py) [reply phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java:3161]: Actually I was of the understanding that I took care of converting existing NULLs in SYSCAT to 0 (default) via: https://github.com/sanjeet006py/phoenix/blob/PHOENIX-7006/phoenix-core/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java#L1448, but I agree it good to be consistent around definition of max lookback age not defined.

**Comment by sanjeet006py** [comment]
Jenkins test this please

File: phoenix-core-client/src/main/java/org/apache/phoenix/schema/TableProperty.java
Comment (gjacoby126) [inline phoenix-core-client/src/main/java/org/apache/phoenix/schema/TableProperty.java:354]: nit: instead of "numeric" specify BIGINT, since "numeric" might refer to a floating point value that isn't supported.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1805]: this can be a constant because it's used at 1830. Also it should be 86400000L to avoid a cast. Alternately you could use Duration.ofDays(1).getSeconds() * 1000 if you want to make it more explicit.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (gjacoby126) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1849]: This would be simpler to do as PhoenixRuntime.getTableNoCache().getMaxLookbackAge()

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1799]: What about if someone sets MAX_LOOKBACK_AGE to NULL?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java
Comment (gjacoby126) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java:1781]: Likewise this would also be simpler using PhoenixRuntime.getTableNoCache() and then looking up from the PTable that's returned.

File: phoenix-core-client/src/main/java/org/apache/phoenix/schema/TableProperty.java
Comment (haridsv) [inline phoenix-core-client/src/main/java/org/apache/phoenix/schema/TableProperty.java:351]: What about -ve numbers?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (haridsv) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1816]: How is this different from above?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (haridsv) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1860]: What about testing with parent table?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java
Comment (haridsv) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java:1730]: This variant is not nneeded, it is really only testing the parser literal token.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java
Comment (haridsv) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java:1775]: As I commented earlier, there is no need to use anything but the simplest table schema here.

**Review by haridsv — CHANGES_REQUESTED** [review]

File: phoenix-core-client/src/main/java/org/apache/phoenix/schema/TableProperty.java
Comment (sanjeet006py) [reply phoenix-core-client/src/main/java/org/apache/phoenix/schema/TableProperty.java:351]: Negative numbers are not parsed by grammar as I tried and got error. As per grammar it seems like only whole numbers are expected.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (sanjeet006py) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1816]: First I created a table with maxlookback age 1 day, then I increased it to 3 days and now decreasing it to 2 days. Lastly setting it to 0.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (sanjeet006py) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1860]: You mean for indexes? If so then it will be part of next batch of changes.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java
Comment (sanjeet006py) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/AlterTableIT.java:1802]: Currently, for most table level properties setting them to NULL via Alter is no-op. Created a JIRA issue: https://issues.apache.org/jira/browse/PHOENIX-7161 for the same. Thanks

**Comment by sanjeet006py** [comment]
@gjacoby126 @haridsv please re-review the PR. Thanks

File: phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java
Comment (kadirozde) [inline phoenix-core-client/src/main/java/org/apache/phoenix/compile/QueryCompiler.java:195]: I do not think we need to get a new ColumnResolver here. Can we use the column resolver of the QueryCompiler object. Please note FromCompiler.getResolverForQuery() makes an RPC call to the syscat region.

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java
Comment (kadirozde) [inline phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:551]: We need to see if the HBase table name for a given Phoenix table has changed by the online data format change feature. We need query the system catalog using WHERE  PHYSICAL_TABLE_NAME = '<this HBase table name>' to find the Phoenix table name.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java
Comment (kadirozde) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/CreateTableIT.java:1798]: Please use conn.getTableNoCache instead.

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java
Comment (sanjeet006py) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:551]: Thanks I missed this scenario. Wondering if same can happen in here also: https://github.com/apache/phoenix/blob/master/phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/UngroupedAggregateRegionObserver.java#L601-L614.

Comment (sanjeet006py) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:551]: If we have a table `T1` and we do a transform on it, seems like SYSCAT will have following two records: 

- A record with `TABLE_NAME` as `T1_1` and `PHYSICAL_TABLE_NAME` also as `T1_1`.
- A record with `TABLE_NAME` as `T1` and `PHYSICAL_TABLE_NAME` as `T1_1`.

As we are looking up Phoenix table from HBase table shall we skip the records where `TABLE_NAME` == `PHYSICAL_TABLE_NAME` and `PHYSICAL_TABLE_NAME is NOT NULL`?

Comment (sanjeet006py) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: As above WHERE clause was causing FULL SCAN over SYSCAT and to avoid that went with an alternate approach. Please take a look. Thanks

Comment (sanjeet006py) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: > What is the alternative approach?

In `TransformClient` we derive new table name from existing table name. I saw we are appending sequence num to eixtsing table's full name separated by underscore. I added a logic to derive logical table name from physical table name by checking if physical table is underscore separated and if part of table name before underscore is valid logical table name in SYSCAT along with checking if the logical table found has physical table as original physical table name.

I thought this will avoid FULL SCAN along max be one RPC call but even any READ operation on any table even if we create an index will be at least one RPC call. Please correct me if I misunderstood something. Thanks

Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: Based on a side conversation with @sanjeet006py, I learned that  the IT tests for Transform is not running currently due to missing @Category annotation and some of these tests are broken. We need to fix the transform tool issues in a separate Jira. We can assume that the transform tool is not supported yet and the name of a Phoenix table and the name of its HBase table are the same for now. When we fix transform tool issues, we need to handle the table name mismatch issues in general.

Comment (kadirozde) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: > I saw we are appending sequence num to eixtsing table's full name separated by underscore
I see. We need to revisit this later to see if it is safe to rely on this.

**Comment by sanjeet006py** [comment]
@kadirozde I have addressed all the review comments. Please re-review. Thanks

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataRegionObserver.java
Comment (virajjasani) [inline phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataRegionObserver.java:64]: unused

**Review by virajjasani — COMMENTED** [review]
Left some comments, will take another look later

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java
Comment (haridsv) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:1629]: There is a recursive call and the resource is not needed for it, so if we close it before the call begins, the resource can be released a bit sooner and so can be beneficial.

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java
Comment (sanjeet006py) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java:550]: Sure, but wondering any drawback with non-parameterized version, just for my knowledge? Thanks

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java
Comment (sanjeet006py) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:1488]: How about `isViewIndex` as we are storing if the PTable instance which we are deserializing is that a view index or not?

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java
Comment (sanjeet006py) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:1649]: Yes, because we need to use `DATA_TABLE_NAME` filed in syscat so needs to handled special.

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java
Comment (sanjeet006py) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/MetaDataEndpointImpl.java:1629]: In this method there are 4 `return` clauses. Having two `close()` calls before each of them seems a bit unclean while using try-with-resources seems clean. If mem by these extra resources is not a major concern then I think try-with-resource might be better choice.

File: phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java
Comment (virajjasani) [reply phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: It's perf improvement depending on the size of the string. When heavy objects are to be logged and if we use non-param version, the string concatenation happens even though the given logger level is not enabled (e.g. even if log level is INFO and the above DEBUG log is not going to be printed, the string concatenation still happens), whereas if we use parameterized version, the string concatenation happens only if the given logger level is enabled and the log is expected to be printed. With parameterized version, the string concatenation takes place within log4j implementation.

**Comment by virajjasani** [comment]
@sanjeet006py the changes overall looks good now on the latest revision. There are some merge conflicts though.

**Comment by sanjeet006py** [comment]
Thanks a lot @virajjasani . I have resolved the conflicts. The conflicts were around the version changes for 5.3.0 creation. Please re-review. Thanks

**Comment by sanjeet006py** [comment]
Once current PR validation is done. Shall I squash and rebase all my changes into one commit or merge procedure will take care of that?

**Review by virajjasani — APPROVED** [review]
+1

**Comment by virajjasani** [comment]
Triggered one more build. Once we get results, will merge the PR unless there are any objections.

**Comment by sanjeet006py** [comment]
@virajjasani the latest build is complete. Are we good to merge the PR? Thanks

**Comment by virajjasani** [comment]
Test failures on the build are not relevant.

**Review by haridsv — COMMENTED** [review]
I had a couple of comments and forgot to submit the review, sorry about doing this after the merge. May be it can help with a future change.

**Pull request merged**
